File: wafel/ui/object_slots.py
from typing import *

from wafel_core import ObjectBehavior

import wafel.imgui as ig


def render_object_slots(
  id: str,
  behaviors: List[Optional[ObjectBehavior]],
  behavior_name: Callable[[ObjectBehavior], str],
) -> Optional[int]:
  ig.push_id(id)

  button_size = 50
  window_left = ig.get_window_position()[0]
  window_right = window_left + ig.get_window_content_region_max()[0]
  prev_item_right = window_left
  style = ig.get_style()

  result = None

  for slot, behavior in enumerate(behaviors):
    item_right = prev_item_right + style.item_spacing[0] + button_size
    if item_right > window_right:
      prev_item_right = window_left
    elif slot != 0:
      ig.same_line()
    prev_item_right = prev_item_right + style.item_spacing[0] + button_size

    if behavior is None:
      label = str(slot)
    else:
      label = str(slot) + '\n' + behavior_name(behavior)

    if ig.button(label + '##slot-' + str(slot), button_size, button_size):
      result = slot

  ig.pop_id()
  return result


__all__ = ['render_object_slots']

File: wafel_app/src/object_slots.rs
use imgui::{self as ig, im_str};
use wafel_core::{ObjectBehavior, ObjectSlot};

pub(crate) fn render_object_slots(
    ui: &ig::Ui<'_>,
    id: &str,
    behaviors: &[Option<ObjectBehavior>],
    behavior_name: impl Fn(&ObjectBehavior) -> String,
) -> Option<ObjectSlot> {
    let id_token = ui.push_id(id);

    let button_size = 50.0;
    let window_left = ui.window_pos()[0];
    let window_right = window_left + ui.window_content_region_max()[0];
    let mut prev_item_right = window_left;
    let style = ui.clone_style();

    let mut result = None;

    for (slot_index, behavior) in behaviors.iter().enumerate() {
        let item_right = prev_item_right + style.item_spacing[0] + button_size;
        if item_right > window_right {
            prev_item_right = window_left;
        } else if slot_index != 0 {
            ui.same_line(0.0);
        }
        prev_item_right = prev_item_right + style.item_spacing[0] + button_size;

        let label = match behavior {
            Some(behavior) => format!("{}\n{}", slot_index, behavior_name(behavior)),
            None => format!("{}", slot_index),
        };

        if ui.button(
            &im_str!("{}##slot-{}", label, slot_index),
            [button_size, button_size],
        ) {
            result = Some(ObjectSlot(slot_index));
        }
    }

    id_token.pop(ui);
    result
}
